### DISCUSSION TITLE
When should I use MPI.Init() & MPI.Finalize() ?

### DISCUSSION THREAD

**Discussion by Roy-Kid** [mpi4py/mpi4py#55] [Q&A] [answered]
Hi, I'm writing a tutorial about mpi4py(but now only the Chinese version available and I will translate it to english soon). I have a question that what does MPI.Init() and Finalize() do? I check the source code in the mpi.pyx and the Init() do invoke the MPI_INIT(), but I don't find where it is invoked when just use `comm = MPI.COMM_WORLD` like the tutorial shows us. Where does the MPI.Init() invoked and should we call it explicitly?

**Comment by dalcinl** [comment, marked as answer]
mpi4py handles MPI initialization and finalization the best possible way it can within the Python environment it runs. You should never ever care about doing it yourself. The only reason `MPI.Init()` and `MPI.Finalize()` are exposed is for these rare cases where users want to handle initialization and finalization manually, perhaps to workaround issues or bugs in third-party C/C++/Fortran MPI code being invoked from Python. 

The way to do manual MPI init/finalize is:

```python
import mpi4py
mpi4py.rc.initialize = False  # do not initialize automatically at import time
mpi4py.rc.finalize = False  # do not finalize automatically at exit

MPI.Init()  # manual initialize

# ...your MPI code ...

MPI.Finalize()  # manual finalize
```

I strongly advice against doing things this way. While manual initialization is not really problematic, manual, premature finalization as show above may trigger issues if your MPI code somehow manages to register cleanup functions (e.g. `__del__` destructors, or functions registered with Python's `atexit` module) making MPI calls. After `MPI.Finalize()`, all these cleanup functions cannot perform MPI calls.

PS: Next time, please submit these kind of questions to our mailing lists at Google Groups, or in the new [GitHub Discussions](https://docs.github.com/en/discussions) section. I'll convert this issue to a GitHub Discussion.
